macaroon.js

```diff
@@ -367,44 +380,64 @@ const Macaroon = class Macaroon {
 };
 
 /**
-  Generates macaroon instances based on the macarono data suppplied.
+  Returns macaroon instances based on the JSON-decoded object
+  in the argument. If this is passed an array, it will decode
+  all the macaroons in the array.
+  TODO accept version 2 format.
   @param {Object|Array} obj A deserialized JSON macaroon or an array of them.
   @return {Macaroon}
 */
-const generateMacaroons = function(obj) {
+const importFromJSONObject = function(obj) {
   if (Array.isArray(obj)) {
-    return obj.map(val => generateMacaroons(val));
+    return obj.map(val => importFromJSONObject(val));
   }
-
-  return new Macaroon({
-    signature: hexToUint8Array(obj.signature),
-    location: obj.location,
-    identifier: obj.identifier,
-    caveats: obj.caveats.map(caveat => {
+  let caveats = []
+  if(obj.caveats !== undefined){
+    caveats = obj.caveats.map(caveat => {
       const _caveat = {
-        _identifier: null,
+        _identifier: requireString(caveat.cid, 'Caveat id.'),
         _location: null,
         _vid: null
       };
-      const cl = caveat.cl;
       if (caveat.cl !== undefined) {
-        requireString(cl, 'Caveat location.');
-        _caveat._location = cl;
+        _caveat._location = requireString(caveat.cl, 'Caveat location.');
```

Comment:
similar if not identical error messages are used in multiple places which might make them hard to find. Maybe make them a little more unique? Or even better would be for them to include information on how to fix the problem and where that problem might be.

Reply:
The error messages are identical to those from before. I think fixing this is a job for another PR. Also, surely the exception stack trace will make them easier to find?